src/util/websocket/WebSocketManager.cpp

```diff
@@ -0,0 +1,241 @@
+    if (potentialUpdate != nullptr) {
+      callback(std::move(potentialUpdate));
+    } else {
+      registerCallback(queryId, webSocketId, std::move(callback));
+    }
+  };
+  return net::async_initiate<CompletionToken, void(TimedClientPayload)>(
+      initiate, token, queryId, webSocketId);
+}
+
+net::awaitable<void> handleClientCommands(websocket& ws,
+                                          WebSocketId webSocketId,
+                                          const QueryId& queryId) {
+  absl::Cleanup cleanup{
+      [webSocketId, queryId]() { disableWebSocket(webSocketId, queryId); }};
+  beast::flat_buffer buffer;
+
+  while (ws.is_open()) {
+    co_await ws.async_read(buffer, net::use_awaitable);
+    // TODO<RobinTF> replace with cancellation process
+    ws.text(ws.got_text());
+    co_await ws.async_write(buffer.data(), net::use_awaitable);
+    buffer.clear();
+  }
+}
+
+net::awaitable<void> waitForServerEvents(websocket& ws, WebSocketId webSocketId,
+                                         const QueryId& queryId) {
+  common::TimeStamp lastUpdate = std::chrono::steady_clock::now();
+  while (ws.is_open()) {
+    auto update = co_await waitForEvent(queryId, webSocketId, lastUpdate,
+                                        net::use_awaitable);
+    if (update == nullptr) {
+      if (ws.is_open()) {
+        co_await ws.async_close(beast::websocket::close_code::normal,
+                                net::use_awaitable);
+      }
+      break;
+    }
+    ws.text(true);
+    const auto& json = update->payload;
+    lastUpdate = update->updateMoment;
+    co_await ws.async_write(net::buffer(json.data(), json.length()),
+                            net::use_awaitable);
+  }
+}
+
+net::awaitable<void> connectionLifecycle(
+    tcp::socket socket, http::request<http::string_body> request) {
+  auto match = ctre::match<"/watch/([^/?]+)">(request.target());
+  AD_CORRECTNESS_CHECK(match);
+  QueryId queryId = QueryId::idFromString(match.get<1>().to_string());
+  websocket ws{std::move(socket)};
+
+  try {
+    ws.set_option(beast::websocket::stream_base::timeout::suggested(
+        beast::role_type::server));
+
+    co_await ws.async_accept(request, boost::asio::use_awaitable);
+
+    WebSocketId webSocketId = WebSocketId::uniqueId();
+    enableWebSocket(webSocketId, queryId);
+    auto strand = net::make_strand(ws.get_executor());
+
+    // experimental operators
+    co_await (
+        net::co_spawn(strand, waitForServerEvents(ws, webSocketId, queryId),
+                      net::use_awaitable) &&
+        net::co_spawn(strand, handleClientCommands(ws, webSocketId, queryId),
+                      net::use_awaitable));
+
+  } catch (boost::system::system_error& error) {
+    if (error.code() == beast::websocket::error::closed) {
+      co_return;
+    }
+    // There was an unexpected error, rethrow
+    throw;
+  }
+
+  // In case an unexpected exception is thrown, close the connection gracefully.
+  if (ws.is_open()) {
```

Comment:
I'm not sure if it is worth trying to catch all kind of exceptions here. If the TCP socket is closed by force the websocket connection will end either way